feat: Adding google calendar pckge

lancemccluskey opened the pull request:
## Description

Adds calendar page to Nodejs.dev, similar to https://calendar.google.com/calendar/u/0/embed?src=nodejs.org_nr77ama8p7d7f9ajrpnu506c98@group.calendar.google.com

## Related Issues

Addresses #883

lancemccluskey commented:
/preview

lancemccluskey commented on src/pages/calendar.tsx:
```diff
@@ -0,0 +1,48 @@
+import Layout from '../components/Layout';
+
+const API_KEY =
+  process.env.GATSBY_REACT_APP_GCAL_API_KEY ||
+  'AIzaSyCBklATFMNjWUjJVZswlTmoyZh27FbaHDQ';
```
What do we typically do for api keys? Mines locked down so i dont mind showing it here

mikeesto commented on package.json:
```diff
@@ -5,6 +5,8 @@
   "author": "Node.js Dev Team <https://github.com/nodejs/nodejs.dev#the-nodejsdev-team>",
   "private": true,
   "dependencies": {
+    "@emotion/react": "^11.4.0",
```
Late last year there was a big effort to [remove emotion](https://github.com/nodejs/nodejs.dev/pull/928) from this project. Perhaps we can use a regular object here instead for styling?

rodion-arr replied:
Agree, looks like calendar component supports regular objects as input param

lancemccluskey commented:
Alright, im pretty confused. I have no idea why its failing here, but everything passes locally. I tried creating a new branch bc I thought that might be it, commenting out the cache step in the test workflow, and even nuking node modules locally and installing with `npm ci`. Does anyone else know why its failing?

lancemccluskey commented:
After digging around, I think its failing because the commit I pushed before was cached so now itll keep failing even though I uninstalled `@emotion/react`. Does anyone know how to bust the cache?

lancemccluskey commented:
Guess not, alright im officially out of ideas

rodion-arr reviewed (commented):
I could reproduce same issue with build locally so it's not related to CI cache.

I assume this is due to [`@emotion/react` is actually a peerDependency of `react-google-calendar`](https://github.com/ericz1803/react-google-calendar/blob/master/package.json#L25)

Locally you are probably using Current node version with npm 7 which installs peer deps. But CI uses node v14 and does not do this by default.

It looks like we need to add `@emotion/react` as dependency anyway :(

lancemccluskey commented:
> I could reproduce same issue with build locally so it's not related to CI cache.
> 
> I assume this is due to [`@emotion/react` is actually a peerDependency of `react-google-calendar`](https://github.com/ericz1803/react-google-calendar/blob/master/package.json#L25)
> 
> Locally you are probably using Current node version with npm 7 which installs peer deps. But CI uses node v14 and does not do this by default.
> 
> It looks like we need to add `@emotion/react` as dependency anyway :(

Ok ill add it back in and push it up. I thought it was weird I kept seeing emotion in my peer deps. Howd you know the npm version wasnt 7 in gh actions? I assumed if it was the same version I had locally it would be fine

benhalverson reviewed (changes requested):
Not ready yet.

lancemccluskey commented:
/preview

manishprivet reviewed (commented):
LGTM

Just a few nitpicks that I have found:

1. Can you make the calendar a little bit smaller, I think some margin around it will make it look more better and concise.
2. Can you make the `Location` links clickable, or any HTTP URL for that matter.
3. There are some minor contrast issues in dark mode.
---

![image](https://user-images.githubusercontent.com/54291836/130307148-5f126bb2-978e-4f00-8929-a021bfa9c299.png)
---
![image](https://user-images.githubusercontent.com/54291836/130307161-edc22839-6d73-486c-b25f-bdd7a840be5d.png)
---
![image](https://user-images.githubusercontent.com/54291836/130307127-dac83190-cfdd-4978-9ccd-5bfa12e21e03.png)

lancemccluskey commented:
> LGTM
> 
> Just a few nitpicks that I have found:
> 
> 1. Can you make the calendar a little bit smaller, I think some margin around it will make it look more better and concise.
> 2. Can you make the `Location` links clickable, or any HTTP URL for that matter.
> 3. There are some minor contrast issues in dark mode.
> 

Yup! Theres still a lot of cleanup to do in the code also. Im hoping to have the changes completed by Sunday/Monday

rodion-arr commented:
Maybe we can mark it as draft so it will be clear when it will be ready for review?

lancemccluskey commented:
> Maybe we can mark it as draft so it will be clear when it will be ready for review?

I would, but I literally just finished lol

benhalverson commented on src/hooks/useGCalAPI.tsx:
```diff
@@ -0,0 +1,52 @@
+      process.env.GATSBY_REACT_APP_GCAL_API_KEY ||
+        'AIzaSyCBklATFMNjWUjJVZswlTmoyZh27FbaHDQ'
+    )
+      .then(() => {
```
Can this be done with async/await I think  our code uses that pattern.

benhalverson commented on src/pages/calendar.tsx:
```diff
@@ -0,0 +1,139 @@
+
+const localizer = momentLocalizer(moment);
+
+const eventCardStyles = {
```
Can these styles be added to calendar.scss?

benhalverson commented on src/hooks/useGCalAPI.tsx:
```diff
@@ -0,0 +1,52 @@
+          setRenderedEvents(initiallyRenderedEvents);
+        });
+      })
+      .catch((err: unknown) => {
```
I think you can use the Error type here.
`.catch((err: Error) =>...`

benhalverson commented on src/pages/calendar.tsx:
```diff
@@ -0,0 +1,139 @@
+        popup
+      />
+      {referenceElement && (
+        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
```
Do we have a todo to make this accessible?

lancemccluskey replied:
Nope ill add it in!

benhalverson commented on src/util/gcalUtils.ts:
```diff
@@ -0,0 +1,236 @@
+  const cancelled: CancelledCalendarEvent[] = [];
+
+  items.forEach(event => {
+    if (event.originalStartTime) {
```
Can you combine these two if statements like this?
`if(event.originalStartTime && event.status === 'cancelled') {`

lancemccluskey replied:
I did it this way because we also have to check for `originalStartTime` and `confirmed` also which indicates a changed calendar event.

benhalverson commented on src/util/gcalUtils.ts:
```diff
@@ -0,0 +1,236 @@
+  betweenStart: Moment,
+  betweenEnd: Moment
+): Date[] {
+  const rstr = `DTSTART:${moment(startTime)
```
what is rstr?

lancemccluskey replied:
its the recurrence rule I believe. Ill spend some time refactoring so its a little easier to follow

benhalverson reviewed (commented):
👍🏻

benhalverson commented:
We shouldn't merge this until the API key is updated to the OpenJS Foundation key.

lancemccluskey commented:
Deleted snapshot test for calendar bc when merging with main discovered the snapshot will change everyday...because its a calendar lol

lancemccluskey commented on test/pages/calendar.test.tsx:
```diff
@@ -0,0 +1,84 @@
+  });
+
+  // TODO: Unable to get tooltip to render
+  it.skip('clicking event shows tooltip', async () => {
```
Super simple test that should pass but isnt. I tried a few different ways to get it working, but for some reason the popper element just isnt rendering. Also noticed the `referenceElement` is not being updated when clicking on the event so might have to dig deeper there

lancemccluskey commented on src/util/gcalUtils.ts:
```diff
@@ -186,7 +187,15 @@ export function getRenderedEvents(
           });
         } else {
-          const eventStart = moment.utc(date);
+          // rrule returns dates in 'UTC' with 0 offset. See https://github.com/jakubroztocil/rrule#important-use-utc-dates
+          const eventStart = moment([
+            date.getUTCFullYear(),
+            date.getUTCMonth(),
+            date.getUTCDate(),
+            date.getUTCHours(),
+            date.getUTCMinutes(),
+            date.getUTCSeconds(),
+          ]);
```
This fixed the issues I was having with timezones being off. I left a comment with a link to where the rrule library talks about it.

manishprivet commented on .env.sample:
```diff
@@ -0,0 +1,4 @@
+# Gatsby requires a prefix of GATSBY and React requires a prefix of REACT_APP
+# Gatsby doc https://www.gatsbyjs.com/docs/how-to/local-development/environment-variables/
+# React doc https://create-react-app.dev/docs/adding-custom-environment-variables/
+GATSBY_REACT_APP_GCAL_API_KEY=AIzaSyCBklATFMNjWUjJVZswlTmoyZh27FbaHDQ
```
Just to confirm, this is not a real key, right?

lancemccluskey replied:
It is a real key. It’s my personal one that needs to be replaced once all the business logic and UI are approved

manishprivet commented on src/config.json:
```diff
@@ -16,5 +16,6 @@
   "facebook": "https://www.facebook.com/nodejsfoundation/",
-  "bannerBtnText": "Blog post"
+  "bannerBtnText": "Blog post",
+  "nodeGcalId": "nodejs.org_nr77ama8p7d7f9ajrpnu506c98@group.calendar.google.com"
```
Can this be hardcoded? @joesepi @MylesBorins

lancemccluskey replied:
It can be. I wasn’t sure if it would change or not so I added it to the config

manishprivet commented on src/util/gcalUtils.ts:
```diff
@@ -0,0 +1,243 @@
+  return events;
+}
+
+// rrule only accepts utc time. https://github.com/jakubroztocil/rrule#important-use-utc-dates
```
```suggestion
// rule only accepts utc time. https://github.com/jakubroztocil/rrule#important-use-utc-dates
```
nit

lancemccluskey replied:
This was on purpose. Rrule references the library which is short for recurrence rule

manishprivet commented on test/pages/calendar.test.tsx:
```diff
@@ -0,0 +1,84 @@
+const thirdStart = moment();
+const fourthStart = moment();
+
+const mockEvents: CalendarEvent[] = [
```
These mock data should be removed, right/

lancemccluskey replied:
No it’s needed for rendering the calendar and asserting in the tests

manishprivet replied:
Oops, sorry missed the file name

manishprivet commented on src/pages/calendar.tsx:
```diff
@@ -0,0 +1,127 @@
+        />
+      </div>
+      {referenceElement && (
+        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
```
Was the accessibility added? If yes I think it can be removed

lancemccluskey replied:
No it wasn’t. I was thinking it could be done in a future PR. I didn’t want the number of changes in one PR to get excessive

joesepi commented:
Great work here. Only thing I see that is a bit of a blocker is the fact that the meeting times are not shown. 🙌

lancemccluskey commented:
> Great work here. Only thing I see that is a bit of a blocker is the fact that the meeting times are not shown. 🙌

Sorry! Don't know how i kept forgetting that. I just added it in now @joesepi

joesepi reviewed (approved):
These are visual nits but I think would be good to have:
- add a little padding to the day boxes so that the text isn't so close to the borders of each day.
- bump up the event title font one or two notches.
- bump up the "title" font (August 2021 in month view for example) to be a bit bigger and perhaps bold.
- Also: I just noticed that on Google Cal proper, the time is bold. That might also be a nice touch.

Otherwise, LGTM!!!!

manishprivet commented:
There are still some minor contrast issues in a dark mode that might be a blocker

![image](https://user-images.githubusercontent.com/54291836/131549927-0b06bd5e-6913-485c-a6a6-2058d46f6f3d.png)

![image](https://user-images.githubusercontent.com/54291836/131549946-95b79e9f-6939-48d1-8d8a-6c0e1531efff.png)
![image](https://user-images.githubusercontent.com/54291836/131549970-0a496942-3138-4c87-889f-b53fa46bdc6e.png)

lancemccluskey commented:
I agree with the nitpicks you mentioned @joesepi and @manishprivet . I was thinking that work could be done in a separate PR though. I imagined this would be a "phase 1" of sorts, and we could iterate upon it. Especially considering how big this Pr is getting.

joesepi commented:
> I agree with the nitpicks you mentioned @joesepi and @manishprivet . I was thinking that work could be done in a separate PR though. I imagined this would be a "phase 1" of sorts, and we could iterate upon it. Especially considering how big this Pr is getting.

Normally, I prefer smaller PRs but this is a big endeavor. I think that if you can address these remaining nits, it would be good to land them in this PR, in my opinion. Thanks!

lancemccluskey commented:
@joesepi @manishprivet I made most of the requested changes detailed below:

* Contrast issue addressed by using color scheme the side nav uses in the learn section
* Increased font size for events on month view
* Increased calendar "title" font size and weight
* Added padding to events in month view so they arent so close to the edges
* (Side Note): Removed hardcoded values from `calendar.scss` and replaced with variables from `tokens.scss`

The only thing I wasnt able to add was making the time bold in the event label for the month view. The reason is because the `titleAccessor` function `react-big-calendar` accepts wants a [string returned](https://jquense.github.io/react-big-calendar/examples/index.html#prop-titleAccessor), not html. We can provide a custom component to the event as described [here](https://jquense.github.io/react-big-calendar/examples/index.html#prop-components) but I really think that should be a separate ticket.

manishprivet commented:
@lancemccluskey The contrast issue still doesn't seem to be fixed here

![image](https://user-images.githubusercontent.com/54291836/132102817-f8c83543-b591-4b57-8b41-090a382f1348.png)

lancemccluskey commented:
@manishprivet What about this? I messed around with the background, but it was hard to find a brand color that looked good on dark and light mode so I went with just changing the font color. I think its more pleasing to the eye and easier to read than before.

<img width="388" alt="Screen Shot 2021-09-04 at 2 34 23 PM" src="https://user-images.githubusercontent.com/40124399/132106031-62e2b807-cf47-4192-b31e-60f2dfceee9d.png">

designMoreWeb commented:
That looks much better

manishprivet commented:
This looks a lot better @lancemccluskey 

![image](https://user-images.githubusercontent.com/54291836/132148584-50e64b80-786b-4162-8269-9c7a675af5c0.png)

Can we do something similar here?

ovflowd commented:
I'm closing this PR as it is stale, and [this](https://github.com/nodejs/nodejs.dev/pull/2428) PR is actively being worked on.
